pkgdb/src/pkgdb/write.cc

```diff
@@ -7,22 +7,293 @@
+      case SR_DISALLOW_RECURSIVE: return "disallowRecursive";
+      default: return "UNKNOWN";
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+void
+RulesTreeNode::addRule( AttrPathGlob & relPath, ScrapeRule rule )
+{
+  /* Modify our rule. */
+  if ( relPath.empty() )
+    {
+      if ( this->rule != SR_DEFAULT )
+        {
+          // TODO: Pass abs-path
+          throw FloxException( "attempted to overwrite existing rule '"
+                               + scrapeRuleToString( this->rule ) + "' for '"
+                               + this->attrName + "' with new rule '"
+                               + scrapeRuleToString( rule ) + "'" );
+        }
+      traceLog( "assigning rule to `" + scrapeRuleToString( rule ) + "' to `"
+                + this->attrName + '\'' );
+      this->rule = rule;
+      return;
+    }
+  traceLog( "adding rule to `" + this->attrName + "': `"
+            + displayableGlobbedPath( relPath ) + " = "
+            + scrapeRuleToString( rule ) + '\'' );
+
+  /* Handle system glob by splitting into 4 recursive calls. */
+  if ( ! relPath.front().has_value() )
+    {
+      traceLog( "splitting system glob into real systems" );
+      for ( const auto & system : getDefaultSystems() )
+        {
+          AttrPathGlob relPathCopy = relPath;
+          relPathCopy.front()      = system;
+          this->addRule( relPathCopy, rule );
+        }
+      return;
+    }
+
+  std::string attrName = std::move( *relPath.front() );
+  // TODO: Use a `std::deque' instead of `std::vector' for efficiency.
+  //       This container is only used for `push_back' and `pop_front'.
+  //       Removing from the front is inefficient for `std::vector'.
+  relPath.erase( relPath.begin() );
+
+  if ( auto it = this->children.find( attrName ); it != this->children.end() )
+    {
+      traceLog( "found existing child `" + attrName + '\'' );
+      /* Add to existing child node. */
+      it->second.addRule( relPath, rule );
+    }
+  else if ( relPath.empty() )
+    {
+      /* Add leaf node. */
+      traceLog( "creating leaf `" + attrName + " = "
+                + scrapeRuleToString( rule ) + '\'' );
+      this->children.emplace( attrName, RulesTreeNode( attrName, rule ) );
+    }
+  else
+    {
+      traceLog( "creating child `" + attrName + '\'' );
+      /* Create a new child node. */
+      this->children.emplace( attrName, RulesTreeNode( attrName ) );
+      this->children.at( attrName ).addRule( relPath, rule );
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+ScrapeRule
+RulesTreeNode::getRule( const AttrPath & path ) const
+{
+  const RulesTreeNode * node = this;
+  for ( const auto & attrName : path )
+    {
+      try
+        {
+          node = &node->children.at( attrName );
+        }
+      catch ( const std::out_of_range & err )
+        {
+          return SR_DEFAULT;
```

Comment:
nit: should this return `SR_NONE` to distinguish between nodes that don't exist and nodes that do?